Qlearning+sarsa v2.0

adrianjav opened the pull request:
This pull request adds more reinforcement learning functionality to dlib. Namely it implements the Qlearning and SARSA algorithms. I think I've done a good job with respect to backward compability and style code so far but let me know anything that needs a review.
A list of the changes:
- Qlearning and SARSA algorithms (files qlearning.h and sarsa.h)
- Separated policy in its own header (policy.h) where there are two implementations (greedy_policy and epsilon_policy).
- An example that shows both algorithms working (qlearning_sarsa_ex.cpp)
- The same example as a unitary test (reinforcement_learning.cpp). It works for me but it'd be nice some more testing since convergence could depend on the machine.
- An interface for the model the agents will work with (model_abstract.h).

About the mechanics: These new algorithms avoid using the process_sample of lspi. Instead they run on a model (usually an online model) that allows them to take steps and get feedback from their choices. What the algorithms really train aren't models but policies. That way you can specify a custom policy and custom initial weights and, when training, the algorithms use an epsilon-greedy policy that uses the given policy as the underlying one.

Related PR: [#824](https://github.com/davisking/dlib/pull/824)

**Edit:**
Mmmm... AppVeyor fails at the unitary tests. Any suggestions on how to test if they work? Since their convergence is non deterministic and qlearning can get worse after converging I find it hard to test.

davisking commented:
Wow, this is a lot of stuff.  Thanks for working on this PR :)

I haven't looked it over very carefully, but it generally looks good.  You should flesh out the comments in the qlearning_sarsa_ex.cpp example a bit so that it reads like a tutorial (e.g. explains the relevant model details so that someone who isn't already familiar with Q learning can understand and learn what it does).  

Also definitely fix the failing unit tests.  Most of them are working though so that's good.   Make sure your random number generators are all seeded the same every run so the tests reliably do the same thing.  You don't want unit tests to have some small chance of failure.  Beyond that I don't have much advice other than to dig in and debug it.

Let me know when you think the PR is ready for a full review.

adrianjav commented:
Thanks god that the tests failed!
Reviewing the code I saw a terrible bug when adding rewards. I've changed that and added custom PRNG support when training. Now everything seems to go smoothly as intended.
I've also rewritten the comments in the example hoping to be clearer.

I think now it's ready for a full review.

travek commented:
Hi!
When this feature will be merged into Master?

davisking commented:
Some time in the next two weeks.  I've been really busy with other stuff
and so haven't had time to review it yet.

davisking commented on dlib/control/approximate_linear_models_abstract.h:
```diff
@@ -56,41 +59,29 @@ namespace dlib
 
-        action_type find_best_action (
-            const state_type& state,
-            const matrix<double,0,1>& w
```
This seems confusing.   I see that the lspi code (in dlib/test/lspi.cpp) still defines feature extractors this way.  So this is wrong isn't it?  This function still needs to be here.

A better way to organize this model/feature extractor stuff is probably to have the documentation say there is just one class, not two.  Then you can define two different versions of the model class.  There is the basic one which was previously defined by the example_feature_extractor.  Then there is the more detailed one with the new things you want to add to models.  The new one has all the functions the feature extractor had, but with some additional new ones.  Then when you use any model class it's not like it depends on some feature extractor class.  It's just one class.

After all, the model is the thing the user makes.  So it's weird that later on it's defined to take a feature extractor as a template argument.

Anyway, my point is that it's really unclear what a model or feature extractor is right now.  Clarifying that is the most important thing here since that's the interface a user of the software is expected to implement.  If it's not really clear people will just not use it and this whole RL module will never be used.  Making that model interface simple so people can easily implement it is the central challenge.

davisking commented on dlib/control/approximate_linear_models_abstract.h:
```diff
@@ -56,41 +59,29 @@ namespace dlib
-            const state_type& state,
-            const action_type& action,
-            matrix<double,0,1>& feats
+        matrix<double,0,1> get_features (
```
This was changed in the new code, but the LSPI code still does it the old way.  Everything needs to be the same.

Also, don't do this.  This change makes the code slower, since now each call to get_features allocates a block of memory.  The other way avoids the reallocation.  I realize move constructors make returning the matrix not super slow.  But reallocating memory in a tight loop is also bad.

davisking commented on dlib/control/approximate_linear_models_abstract.h:
```diff
@@ -128,86 +119,5 @@ namespace dlib
-    template <
-        typename feature_extractor
-        >
-    class policy
```
I don't really like that approximate_linear_models_abstract.h is so empty now.  The general idea with this file is that someone can read it and understand how to use the module.  It defines the key interfaces they need to implement (i.e. the model/feature extractor class the user needs to make).  It also shows them where it goes in the policy object.  Breaking it up into too many files just makes it hard for people to follow the thread.  

So put the policy objects in this file, the same with the documentation about what exactly constitutes the model interface.

davisking commented on dlib/control/model_abstract.h:
```diff
@@ -0,0 +1,169 @@
+{
+
+    template <
+            template<typename, typename> class feature_extractor_type
```
Why is this template argument here?

davisking commented on dlib/control/model_abstract.h:
```diff
@@ -0,0 +1,169 @@
+        !*/
+
+        auto states_size(
+        ) const -> decltype(get_feature_extractor().num_features());
```
Don't use decltype here.  It makes the documentation hard to read.  Should a user really have to go look up the feature extractor interface's num_features to find out that state_size will return an unsigned long?

davisking replied:
In general, don't use decltype in the documentation/abstract files unless it's needed.  Like if the return type is determined by some complex template output or something like that.  And even in those cases it's probably better to have a typedef in the class and to return that.

I guess my point is that the purpose of the stuff in the abstract file is to be documentation that's easily read by a user.  So writing it to be clear is the goal.  That's different from writing code where normally trying not to repeat yourself is a good idea, usually.  But in documentation, you don't want to write in exactly one place that something is an unsigned long and then constantly make people refer back to that to find out that yes, it's an unsigned long.

davisking commented on dlib/control/model_abstract.h:
```diff
@@ -0,0 +1,169 @@
+                - returns get_feature_extractor().num_features().
+        !*/
+
+        auto get_features(
```
Don't allocate and return a matrix every time, per the other comment.

davisking commented on dlib/control/policy.h:
```diff
@@ -0,0 +1,172 @@
+    template <
+        typename model_type
+        >
+    class greedy_policy
```
Rename this to policy rather than having the alias statement later on.  This is the default policy people will use in most cases right?

davisking replied:
Also, put the policy objects back into approximate_linear_models.h.  Save moving things around to another PR since moving stuff around makes the review much harder.  I can't tell what things are just moves vs substantive code changes.

Also, per my other comment, I'm not sure it's a good idea to move these things outside approximate_linear_models_abstract.h anyway.

davisking commented on dlib/control/policy_abstract.h:
```diff
@@ -0,0 +1,284 @@
+#include <random>
+
+namespace dlib
+{
```
Always use correct tabbing.

davisking commented on dlib/control/policy_abstract.h:
```diff
@@ -0,0 +1,284 @@
+template <
+    typename model_type
+    >
+class example_policy
```
Why is there an example policy object?

davisking commented on dlib/control/qlearning_abstract.h:
```diff
@@ -0,0 +1,205 @@
+                  model_abstract.h.
+
+            WHAT THIS OBJECT REPRESENTS
+                This objects is an implementation of the well-known reinforcement learning
```
Fix spelling.

davisking replied:
object**s** and algorithm**s**

davisking replied:
The text is wrong, it doesn't use any process_samples

Please make sure all the documentation is correct.  I am extremely anal about having correct and clear documentation :).  You should review all the comments a few times to make sure they are correct and clear before the next PR review.

davisking commented on dlib/control/qlearning_abstract.h:
```diff
@@ -0,0 +1,205 @@
+        ) const;
+        /*!
+            requires
+                - policy is of the form example_policy<model_type>, i.e., an instance of
```
This is too general.  You should change this function to take a `policy<model>` which is much more specific.  So far I don't see any reason to not be more specific.  That is, why would a user want to create their own policy object?

davisking commented on dlib/control/qlearning_abstract.h:
```diff
@@ -0,0 +1,205 @@
+                  defined in std::random. By default it assumes it to be the standard
+                  default_random_engine class.
+            ensures
+                - returns a policy of the type policy_type as the result of applying the
```
The wording of this is confusing and ungrammatical.

davisking commented on dlib/control/sarsa.h:
```diff
@@ -0,0 +1,171 @@
+            );
+
+            reward_type total_reward = static_cast<reward_type>(0);
+            for(auto iter = 0u; iter < iterations; ++iter){
```
Put { on a line by itself so the code style is consistent with all the other code in this file and other parts of dlib.

adrianjav commented:
Thanks for your comments!

I strongly agree with you in merging feature_extractor and model in one class. I might have overthought it just because feature_extractor was already there and I tried to separate the model and its representation. But at the end it's confusing for the user.

The other things come mostly from: rushing; being used to my own code style; English not being my first language. 

I'll try to fix them all in the next few days.

adrianjav commented:
It's been a month since my last post and I feel like I have to say something. I've been applying the changes of the review but I'm still missing a few things.

I haven't forgotten it, just being really busy with my actual job. As soon as I have some spare time I'll post the last changes

adrianjav commented:
Finally I've applied those requested changes into the old code. The most remarkable change is that, as you suggested, I describe in the abstract two classes: offline model (the former feature extractor) and online model; So the user can choose to implement one of them depending on its necessities.

Overall I agree with the notes you took in the review, the only thing I could argue is that I returned matrices by value on purpose trusting the compiler's ROV optimization. Either way I've also changed that and reviewed the comments twice.

davisking commented:
Cool, thanks for the update.  I'll review the PR sometime over the coming week.

davisking commented:
I haven't forgotten about this PR.  But I've got a new baby at home and
have been a little busy :).  I'll get to it though.

adrianjav commented:
Hahaha don't worry, I'm sure you'll do it. I'm quite busy as well.
Congratulations for your newborn!  :baby:

davisking commented:
Thanks :)

davisking commented:
Not sure.  I'll get to it thought.  You can use it now though, no need to wait on me.  In fact, other people testing and reporting back here will help me in my own review when I get to it.

adrianjav commented:
Thanks for another good review Davis! I've just looked quickly the comments and I agree with what I've read so far. I'm quite busy these days but 'll try to fix everything in the next weeks

davisking commented:
No problem :)

adrianjav commented on dlib/control/qlearning_abstract.h:
```diff
@@ -0,0 +1,182 @@
+        );
+        /*!
+            requires
+                - discount >= 0 and discount <= 1.
```
I think so, a discount of 0 means that the algorithm is totally short-sighted. It will only consider the immediate reward.

adrianjav commented on dlib/control/approximate_linear_models.h:
```diff
@@ -56,68 +56,147 @@ namespace dlib
+        ) const
+        {
+            std::bernoulli_distribution d(epsilon);
+            return d(gen) ? get_model().random_action(state) : underlying_policy(state);
```
random action is a method from online model, I have sustituted that for 
```
return d(gen) ? underlying_policy.get_model().random_action(state) : underlying_policy(state);
```

adrianjav commented:
I think I've fixed everything so far. Some things that are worth mentioning (skiping documentation changes):
* I have removed the thread-safety requirements. It doesn't that useful for those classes and "keep it simple, stupid".
* I'm keeping the mutable random_engine attribute on the classes. I don't consider it a internal state of the object and the only reason is not static/global is because of serialization.
* Serialization works. I have added a serialization test to see whether it works as expected.
* For keeping the code clean, I have added to `serialize.h` methods for all the instances of random devices from the standard library. They are serialized by just dumping their internal code into streams.

Waiting for travis and appveyor tests to finish though.

sandsmark commented:
Are all the comments from the previous review fixed? Just in case Someone™ wanted to pick this up to get it merged.

davisking commented:
I'm not sure. I had my second child right around the time this PR was made, and got severely sidetracked by that and a bunch of work stuff.  I never ended up looking at this PR, since it's big and time consuming to review.  I admit I feel bad about this, and I've left the PR open as a reminder of my shame in reviewing this.  One of the issues though is I'm not sure if anyone really wants to use this.  The state-of-the-art in RL is all deep learning stuff, rather than the simpler things in this diff.  So I'm hesitant to merge things that are not really useful to users (that's not to say @adrianjav didn't do a nice job.  He absolutely did.)

Anyway, are you @sandsmark or @adrianjav using this stuff or interested in it?  If so then maybe it's worth it. In particular, is @adrianjav using this and getting value out of it?

adrianjav commented:
Hi @sandsmark. This PR is two-year old, as far as I remember I fixed everything that was reviewed, although I am not sure anymore.

First, thanks @davisking :) Regarding your questions, I am not using nor interested in RL stuff. This was an attempt of making my university exercises useful for a broader audience and I found a sweet spot in your library. At this moment I am doing a PhD in machine learning, yet I don't focus on RL at all, so I don't get any real value out of this.

With that said, I believe that having these (simple) implementations in your library can't be harmful. Worst case it will be used for educational purposes or as a template to implement more complex algorithms (e.g., PPO). I will be happy to rescue this PR if that will useful for someone.

davisking commented:
Yeah, I guess at least it's useful as an educational resource.  I'll go over this again this weekend and see about merging it :)

davisking commented:
Can you give me write access to this PR?  See for instructions https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

I just rebased this branch on master, fixed some conflicts, and tried to push to the PR but it's not letting me.

adrianjav commented:
The "Allow edits from maintainers" option was already selected for me, you should have write access to my branch.

davisking commented:
Huh.  I just tried again and still the same error.
```
remote: Resolving deltas: 100% (108/108), completed with 11 local objects.
To https://github.com/adrianjav/dlib.git
 ! [remote rejected]   adrianjav-qlearning+sarsa -> adrianjav-qlearning+sarsa (permission denied)
error: failed to push some refs to 'https://github.com/adrianjav/dlib.git'
```
Anyway, maybe it's because I'm trying to push a rebased change.  Can you rebase on master and resolve the errors?

adrianjav commented:
Ok, that is weird, this is what I actually see from my side:
![image](https://user-images.githubusercontent.com/25713831/75634677-0c129d00-5c10-11ea-98c8-e45c22c4371b.png)

I will try to rebase to my master branch tomorrow, I don't have the git in local at the moment. I will be back to you as soon as I do it.

adrianjav commented:
I just rebased this branch with the latest version of dlib. I locally tested the example and unit tests and both work. Hope you can edit the branch now.

sandsmark commented:
> Anyway, are you @sandsmark or @adrianjav using this stuff or interested in it?

For some definition of "use", I guess. I used to make a game a year for an AI competition I held, and I wanted to test SARSA with them. Just for fun, and because I've been thinking about starting the competition up again, and it's much easier to tune games before release with a non-trivial bot (and so far I only used my own trivial Q-learning implementations).

And the reason I asked was because I thought about fixing any remaining issues myself to get it merged, but I wasn't entirely sure what needed to be fixed.

davisking commented:
I'll go over this PR again and probably fix whatever needs fixing (if anything) and merge it.  It's a big PR though so it needs a chunk of time where I can sit and work on it.  I just haven't been able to get time for it the last few weeks.  I'll do it soon though.

adrianjav commented:
Totally understandable, these last weeks haven't been easy for anyone. Stay safe!
